tests/engine/training/test_graph_trainer.py

```diff
@@ -0,0 +1,236 @@
+                config={"test_value": test_value,},
+                is_target=True,
+            ),
+            "load": SchemaNode(
+                needs={"resource": "train"},
+                uses=PersistableTestComponent,
+                fn="run_inference",
+                constructor_name="load",
+                config={},
+            ),
+        }
+    )
+    predict_schema = GraphSchema(
+        {
+            "load": SchemaNode(
+                needs={},
+                uses=PersistableTestComponent,
+                fn="run_inference",
+                constructor_name="load",
+                config={},
+                is_target=True,
+                resource=Resource("train"),
+            ),
+        }
+    )
+
+    output_filename = tmp_path / "model.tar.gz"
+    predict_graph_runner = graph_trainer.train(
+        train_schema=train_schema,
+        predict_schema=predict_schema,
+        domain_path=domain_path,
+        output_filename=output_filename,
+    )
+    assert isinstance(predict_graph_runner, DaskGraphRunner)
+    assert output_filename.is_file()
+    assert predict_graph_runner.run() == {"load": test_value}
+
+
+def test_graph_trainer_fingerprints_and_caches(
+    temp_cache: TrainingCache,
+    tmp_path: Path,
+    train_with_schema: Callable,
+    spy_on_all_components: Callable,
+):
+
+    input_file = tmp_path / "input_file.txt"
+    input_file.write_text("3")
+
+    train_schema = GraphSchema(
+        {
+            "read_file": SchemaNode(
+                needs={},
+                uses=FileReader,
+                fn="read",
+                constructor_name="create",
+                config={"file_path": str(input_file)},
+                is_input=True,
+            ),
+            "train": SchemaNode(
+                needs={},
+                uses=PersistableTestComponent,
+                fn="train",
+                constructor_name="create",
+                config={"test_value": "4"},
+                is_target=True,
+            ),
+            "process": SchemaNode(
+                needs={"resource": "train"},
+                uses=PersistableTestComponent,
+                fn="run_inference",
+                constructor_name="load",
+                config={"wrap_output_in_cacheable": True},
+            ),
+            "add": SchemaNode(
+                needs={"i1": "read_file", "i2": "process"},
+                uses=AddInputs,
+                fn="add",
+                constructor_name="create",
+                config={},
+            ),
+            "assert_node": SchemaNode(
+                needs={"i": "add"},
+                uses=AssertComponent,
+                fn="run_assert",
+                constructor_name="create",
+                config={"value_to_assert": 7},
+                is_target=True,
+            ),
+        }
+    )
+
+    # The first train should call all the components
+    mocks = spy_on_all_components(train_schema)
```

Comment:
how about moving the `spy_on_all_components` to the top of the module?

Reply:
This call? Or the fixture?

Reply:
the fixture 👍🏻